sqlglot/optimizer/qualify_columns.py

```diff
@@ -0,0 +1,321 @@
+    for table in tables:
+        table_name = table.text("this")
+
+        if isinstance(table.parent, exp.Alias):
+            selectable_table = table.parent.alias
+        else:
+            selectable_table = table_name
+
+        if table_name in parent_selectables:
+            selectable_columns = parent_selectables[table_name]
+        else:
+            selectable_columns = list(schema[table_name])
+
+        selectables = _merge(selectables, {selectable_table: selectable_columns})
+    return selectables
+
+
+def _expand_stars(select_stars, selectables):
+    result = []
+    for star in select_stars:
+        new_columns = []
+
+        if isinstance(star, exp.Column):
+            tables = [star.text("table")]
+        else:
+            tables = list(selectables)
+
+        for table in tables:
+            columns = selectables.get(table, [])
+            for column in sorted(columns):
+                new_columns.append(
+                    exp.Column(
+                        this=exp.to_identifier(column), table=exp.to_identifier(table)
+                    )
+                )
+        star.replace(*new_columns)
+        result.extend(new_columns)
+    return result
+
+
+def _qualify_columns(columns, selectables):
+    unambiguous_columns = None  # lazily loaded
+
+    for column in columns:
+        column_table = column.text("table")
+
+        if column_table and column_table not in selectables:
+            raise RuntimeError(f"Unknown table reference: {column_table}")
+        if not column_table:
+            if unambiguous_columns is None:
+                unambiguous_columns = _get_unambiguous_columns(selectables)
+
+            column_table = unambiguous_columns.get(column.text("this"))
+            if not column_table:
+                raise RuntimeError(f"Ambiguous column: {column}")
+            column.set("table", exp.to_identifier(column_table))
+
+
+def _qualify_subqueries(subqueries, schema, sequence, selectables):
+    for subquery in subqueries:
+        _qualify_union(subquery, schema, sequence, selectables)
+
+
+def _qualify_outputs(selections):
+    outputs = []
+
+    for i, selection in enumerate(selections):
+        if isinstance(selection, exp.Alias):
+            selection_name = selection.text("alias")
+        elif isinstance(selection, exp.Column):
+            selection_name = selection.text("this")
+            selection.replace(exp.alias_(selection.copy(), selection_name))
+        else:
+            selection_name = f"_col{i}"
+            selection.replace(exp.alias_(selection.copy(), selection_name))
+
+        outputs.append(selection_name)
+
+    return outputs
+
+
+def _merge(*selectables):
+    """
+    Merge two "selectable" dictionaries.
+
+    Raise an error if there are any duplicate selectable names.
+    """
+    result = {}
+    for s in selectables:
+        for selectable_table, selectable_columns in s.items():
+            if selectable_table in selectables:
```

Comment:
is this correct? selectables in a tuple right, so this will never be true, did you mean in result instead?

Reply:
Oo wow nice catch. 
I haven't added the "invalid" tests yet. Hopefully those would've caught this.